nbconvert/preprocessors/execute.py

```diff
@@ -580,67 +580,45 @@ def run_cell(self, cell, cell_index=0, store_history=False):
-
-        while more_output or polling_exec_reply:
-            if polling_exec_reply:
-                if self._passed_deadline(deadline):
-                    self._handle_timeout(exec_timeout, cell)
-                    polling_exec_reply = False
-                    continue
-
-                # Avoid exceeding the execution timeout (deadline), but stop
-                # after at most 1s so we can poll output from iopub_channel.
-                timeout = self._timeout_with_deadline(1, deadline)
-                exec_reply = self._poll_for_reply(parent_msg_id, cell, timeout)
-                if exec_reply is not None:
-                    polling_exec_reply = False
-
-            if more_output:
-                try:
-                    timeout = self.iopub_timeout
-                    if polling_exec_reply:
-                        # Avoid exceeding the execution timeout (deadline) while
-                        # polling for output.
-                        timeout = self._timeout_with_deadline(timeout, deadline)
-                    msg = self.kc.iopub_channel.get_msg(timeout=timeout)
-                except Empty:
-                    if polling_exec_reply:
-                        # Still waiting for execution to finish so we expect that
-                        # output may not always be produced yet.
-                        continue
-
-                    if self.raise_on_iopub_timeout:
-                        raise CellTimeoutError.error_from_timeout_and_cell(
-                            "Timeout waiting for IOPub output", self.iopub_timeout, cell
-                        )
+        # we need to have a reply, and return to idle before we can consider the cell executed
+        idle = False
+        execute_reply = None
+        while not idle or execute_reply is None:
+            # we want to timeout regularly, to see if the kernel is still alive
+            # this is tested in preprocessors/test/test_execute.py#test_kernel_death
+            # this actually fakes the kernel death, and we might be able to use the xlist
+            # to detect a disconnected kernel
+            timeout = min(1, deadline - monotonic())
+            # if we interrupt on timeout, we allow 1 seconds to pass till deadline
+            # to make sure we get the interrupt message
+            if timeout >= (-1 if self.interrupt_on_timeout else 0):
+                # we include 0, which simply is a poll to see if we have messages left
+                rlist, wlist, xlist = zmq.select([self.kc.iopub_channel.socket, self.kc.shell_channel.socket], [], [], min(0, timeout))
+                # print("lists", rlist, wlist, xlist)
+                if not rlist and not wlist and not xlist:
+                    self._check_alive()
+                    if monotonic() > deadline:
+                        self._handle_timeout(exec_timeout, cell)
+                if xlist:
+                    raise RuntimeError("Oops, unexpected rror")
```

Comment:
```suggestion
                    raise RuntimeError("Oops, unexpected error")
```

Reply:
(actually, the whole error message isn't that helpful :)

Reply:
Yeah, this is partly the reason this is a draft, we don't have a test that could trigger this. My guess is that maybe closing the socket from the kernel side might do it, or segfaulting the kernel.